### PR TITLE
BUG: fix compilation error for VS 141 and earlier

### LINKED ISSUE #21346: BUG: error compiling on MSVC 141 toolchain
### Describe the issue:

When compiling current Numpy main (ae1330773) on MSVC v141, with `pip install -v -v .`, I get the following compilation error:

```
C:\repos\numpy\numpy\core\src\npysort\x86-qsort.dispatch.cpp(651): error C3861: '_knot_mask16': identifier not found
```

This is with:

```
Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27045 for x64
```

If I'm using the 142 toolchain, Numpy compiles to completion:

```
Microsoft (R) C/C++ Optimizing Compiler Version 19.29.30139 for x64
```

In both cases:

```
Python 3.9.6 (tags/v3.9.6:db3ff76, Jun 28 2021, 15:26:21) [MSC v.1929 64 bit (AMD64)] on win32
```

### Reproduce the code example:

```python
See above.
```

### Error message:

```shell
See above.
```

### NumPy/Python version information:

See above.

### PR DESCRIPTION
**Pull request by mattip**
Compilers before vs 1922 are missing the `_knot_mask16` intrinsic, use a simple `NOT` instead.

Also clean up a few stray `npy_intp`/`npy_int` warnings.

Fixes #21346

I verified the compiler version using https://godbolt.org/z/KMYhzaM1e.

### CODE REVIEW

**Comment by mattip** [comment]
The CI runs VS 2019 and will not test the other `#ifdef` branch. Testing this requires a AVX512 machine + building with [@matthew-brett's formula](https://github.com/numpy/numpy/issues/21346#issuecomment-1102862089) to set the compiler chain to 141:
```
"C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Auxiliary\Build\vcvarsall.bat" x64 -vcvars_ver=14.16
```

I had to `s/BuildTools/Community` and use the Visual Studio installer to install the 14.16 toolchain.

**Comment by serge-sans-paille** [comment]
Looks good to me! https://godbolt.org/z/YohW1Tdd8

**Comment by charris** [comment]
Close/reopen

File: numpy/core/src/npysort/x86-qsort.dispatch.cpp [hunk 1]
```diff
@@ -808,7 +812,7 @@ NPY_CPU_DISPATCH_CURFX(x86_quicksort_int)(void *arr, npy_intp arrsize)
 {
     if (arrsize > 1) {
         qsort_<vector<npy_int>, npy_int>((npy_int *)arr, 0, arrsize - 1,
-                                         2 * log2(arrsize));
+                                         2 * (npy_int)log2(arrsize));
```
Comment (charris) [inline numpy/core/src/npysort/x86-qsort.dispatch.cpp:815]: Hmm. I don't know if this even matters much, but could do  `(npy_int)(2 * log2(arrsize)));`

**Comment by charris** [comment]
Let's give this a shot. I do worry about the complexity of this implementation of quicksort, but that is a bigger problem.

Thanks Matti.

**Pull request merged**